src/hashupdater.cpp

```diff
@@ -0,0 +1,100 @@
+#include "hashupdater.hpp"
+#include "util.hpp"
+#include "world.hpp"
+
+PasswordHashUpdater::PasswordHashUpdater(Config& config, const std::unordered_map<HashFunc, std::shared_ptr<Hasher>>& passwordHashers)
+    : _config(config)
+    , _passwordHashers(passwordHashers)
+    , _database(nullptr)
+    , _terminating(false)
+{
+    this->_updateThread = std::thread([this] { this->updateThreadProc(); });
+
+    auto dbType = util::lowercase(std::string(config["DBType"]));
+
+    Database::Engine engine = Database::MySQL;
+    if (!dbType.compare("sqlite"))
+    {
+        engine = Database::SQLite;
+    }
+    else if (!dbType.compare("sqlserver"))
+    {
+        engine = Database::SqlServer;
+    }
+    else if (!dbType.compare("mysql"))
+    {
+        engine = Database::MySQL;
+    }
+
+    auto dbHost = std::string(config["DBHost"]);
+    auto dbUser = std::string(config["DBUser"]);
+    auto dbPass = std::string(config["DBPass"]);
+    auto dbName = std::string(config["DBName"]);
+    auto dbPort = int(config["DBPort"]);
+
+    this->_database.reset(new Database(engine, dbHost, dbPort, dbUser, dbPass, dbName));
+}
+
+PasswordHashUpdater::~PasswordHashUpdater()
+{
+    this->_terminating = true;
+
+    this->_updateSignal.notify_all();
+    this->_updateThread.join();
+}
+
+void PasswordHashUpdater::QueueUpdatePassword(const std::string& username, util::secure_string&& password, HashFunc hashFunc)
+{
+    UpdateState state { username, std::move(password), hashFunc };
+
+    std::lock_guard<std::mutex> queueGuard(this->_updateQueueLock);
+    this->_updateQueue.push(std::move(state));
+
+    this->_updateSignal.notify_one();
+}
+
+void PasswordHashUpdater::updateThreadProc()
+{
+    while (!this->_terminating)
+    {
+        std::unique_lock<std::mutex> lock(this->_updateMutex);
+        this->_updateSignal.wait(lock);
```

Comment:
Technically, it looks like it would be possible for `PasswordHashUpdater::QueueUpdatePassword` to be called twice before `PasswordHashUpdater::updateThreadProc` finishes. I think that could lead to the following scenario:

```txt
queue: {}
T2: waiting ...

T1: queue A
queue: {A}
T1: notify
T2: process A - begin
queue: {}

T1: queue B
queue: {B}
T1: notify
T2: process A - end

queue: {B}
T2: waiting ...
```

It's late here and I may have overlooked a part of the logic, so feel free to let me know if I'm wrong.

Reply:
That's a valid concern, I had the same thought. I added a lock around the queue access for just this purpose (`this->updateQueueLock`). I think the way it's written, the second notify would cause the next wait to immediately return and process B, so nothing gets "stuck" like in your example.

Could you double-check the logic and see if that makes sense? I don't want to be confusing the behavior of `std::condition_variable` with how `ManualResetEvent` works in C#.

Reply:
I think the problem here is that the update queue lock is only held by T1 until it notifies T2, at which point it goes out of scope and is released. I'll annotate my example above to indicate where the lock could be released that would be problematic:

```txt
queue: {}
T2: waiting ...

T1: queue A <- obtains updateQueueLock
queue: {A}
T1: notify
T2: process A - begin <- obtains updateMutex
queue: {}
T1: exits QueueUpdatePassword scope <- releases updateQueueLock

T1: queue B <- obtains updateQueueLock
queue: {B}
T1: notify
T2: process A - end <- releases updateMutex
T1: exits QueueUpdatePassword scope <- releases updateQueueLock

queue: {B}
T2: waiting ...
```

Reply:
Let me know if you see a gap in my logic; if I was always 100% correct I'd be flying around in a private jet by now.

Reply:
@iandinwoodie this concern should be addressed now, let me know if you see issues with the new implementation